### ISSUE TITLE
Size is missing from HotColdRegions when started with Visual Studio Debugger

### ISSUE THREAD

**Issue by PathogenDavid** [microsoft/clrmd#869]
If the target process was started by the Visual Studio Debugger, `HotColdRegions.HotSize` will be 0. Additionally, `ClrMethod.CompilationType` will be `None`.

This is occurring with .NET Core 3.1 as well as .NET 5 on Windows 10 20H2 with Visual Studio 2019 16.8.1

This does not occur if attaching to a process that was started without debugging. It also doesn't occur if you start the process without debugging and then attach the Visual Studio Debugger after it starts. (Enable the loop at the beginning of `Main`)

Here's a reduced repro:

```csharp
using Microsoft.Diagnostics.Runtime;
using System;
using System.Diagnostics;
using System.Linq;
using System.Runtime.CompilerServices;

public static class Program
{
    public static int Add(int a, int b)
    {
        return a + b;
    }

    public static void Main()
    {
#if false
        Console.WriteLine("Waiting for debugger...");
        while (!Debugger.IsAttached)
        { }
#endif

        // Get the method and make sure it's JITted
        RuntimeMethodHandle methodHandle = typeof(Program).GetMethod(nameof(Add)).MethodHandle;
        RuntimeHelpers.PrepareMethod(methodHandle);

        // Attach ClrMd
        using DataTarget dt = DataTarget.CreateSnapshotAndAttach(Process.GetCurrentProcess().Id);

        ClrInfo runtimeInfo = dt.ClrVersions.First();
        ClrRuntime runtime = runtimeInfo.CreateRuntime();
        ClrMethod method = runtime.GetMethodByHandle((ulong)methodHandle.Value);

        Console.WriteLine($"Method: {method.Signature}");
        Console.WriteLine($"Compilation type: {method.CompilationType}");
        Console.WriteLine($"Method start: 0x{method.NativeCode:X16}");
        HotColdRegions hotCold = method.HotColdInfo;
        Console.WriteLine($"Cold region: 0x{hotCold.ColdStart:X16} for {hotCold.ColdSize}");
        Console.WriteLine($"Hot region: 0x{hotCold.HotStart:X16} for {hotCold.HotSize}");

        Console.WriteLine("Done.");
    }
}
```

Here is the output of this program in various scenarios:

<table>
<tr><th>With debugging (F5)</th><th>Without debugging (Ctrl+F5)</th></tr>
<tr>
<td>

```
Method: Program.Add(Int32, Int32)
Compilation type: None
Method start: 0x00007FFCCC798700
Cold region: 0x0000000000000000 for 0
Hot region: 0x00007FFCCC798700 for 0
Done.
```

</td><td>

```
Method: Program.Add(Int32, Int32)
Compilation type: Jit
Method start: 0x00007FFCCCA94CA0
Cold region: 0x0000000000000000 for 0
Hot region: 0x00007FFCCCA94CA0 for 57
Done.
```

</td></tr></table>

**Comment by PathogenDavid** [comment]
.NET Framework 4.8 is a similar situation, but not quite the size is 6 instead of 0:

<table>
<tr><th>With debugging (F5)</th><th>Without debugging (Ctrl+F5)</th></tr>
<tr>
<td>

```
Method: Program.Add(Int32, Int32)
Compilation type: None
Method start: 0x00007FFCF7AC0DD0
Cold region: 0x0000000000000000 for 0
Hot region: 0x00007FFCF7AC0DD0 for 6
Done.
```

</td><td>

```
Method: Program.Add(Int32, Int32)
Compilation type: Jit
Method start: 0x00007FFCF7AE0DD0
Cold region: 0x0000000000000000 for 0
Hot region: 0x00007FFCF7AE0DD0 for 57
Done.
```

</td></tr></table>

**Comment by PathogenDavid** [comment]
Perhaps related, .NET 5 actually isn't working at all regardless of Visual Studio attachment:

<table>
<tr><th>With debugging (F5)</th><th>Without debugging (Ctrl+F5)</th></tr>
<tr>
<td>

```
Method: Program.Add(Int32, Int32)
Compilation type: None
Method start: 0x00007FF9CFBAA320
Cold region: 0x0000000000000000 for 0
Hot region: 0x00007FF9CFBAA320 for 0
Done.
```

</td><td>

```
Method: Program.Add(Int32, Int32)
Compilation type: None
Method start: 0x00007FF9CFE2A2E0
Cold region: 0x0000000000000000 for 0
Hot region: 0x00007FF9CFE2A2E0 for 0
Done.
```

</td></tr></table>

**Comment by leculver** [comment]
@PathogenDavid I can take a look next time I do a sprint of clrmd work, but this is likely either: 

1)  A real change within the CLR Runtime, where VS puts the process into a unique state when debugging/not debugging/etc.
2)  A bug within the dac (the underlying diagnostic library that ClrMD calls into).

You could confirm that second case by checking the SOS output of `!dumpmt -md [ClrType.MethodTable]` and checking the compilation type it reports.  I'll take a look next time I'm doing some ClrMD related work.  I'm mostly on vacation this week (due to thanksgiving).

**Comment by PathogenDavid** [comment]
Thanks for the info! And no worries, hope you enjoy your time off!

Here's the output from the program and from WinDbg. This is .NET 5 launched from WinDbg.

```
Method: Program.Add(Int32, Int32)
Compilation type: None
Method start: 0x00007FF9A066A2E0
Cold region: 0x0000000000000000 for 0
Hot region: 0x00007FF9A066A2E0 for 0
Done.
```

```
0:009> !dumpmt -md 00007ff9a0704958
EEClass:         00007ff9a071a940
Module:          00007ff9a07027e0
Name:            Program
mdToken:         0000000002000002
File:            C:\Development\Playground\ClrMdDumpJit\PrintMethodInfo\bin\Debug\net5.0\PrintMethodInfo.dll
BaseSize:        0x18
ComponentSize:   0x0
Slots in VTable: 6
Number of IFaces in IFaceMap: 0
--------------------------------------
MethodDesc Table
           Entry       MethodDesc    JIT Name
00007FF9A0650090 00007ff9a0620bf8   NONE System.Object.Finalize()
00007FF9A0650098 00007ff9a0620c08   NONE System.Object.ToString()
00007FF9A06500A0 00007ff9a0620c18   NONE System.Object.Equals(System.Object)
00007FF9A06500B8 00007ff9a0620c58   NONE System.Object.GetHashCode()
00007FF9A065C758 00007ff9a0704928    JIT Program.Add(Int32, Int32)
00007FF9A065C760 00007ff9a0704940    JIT Program.Main()
```

(I'm not super experience with WinDbg so I stumbled my way to that `00007ff9a0704958` argument using `!name2ee *!Program`. If that wasn't what you meant let me know.)

-----------

<details>
<summary>I realized I never checked if it happened with .NET Core 3.1 launched from WinDbg, so here's that too (Click to expand)</summary>

```
Method: Program.Add(Int32, Int32)
Compilation type: Jit
Method start: 0x00007FF99A094CB0
Cold region: 0x0000000000000000 for 0
Hot region: 0x00007FF99A094CB0 for 57
Done.
```

```
0:010> !dumpmt -md 00007ff99a131b80
EEClass:         00007ff99a146e60
Module:          00007ff99a12f808
Name:            Program
mdToken:         0000000002000002
File:            C:\Development\Playground\ClrMdDumpJit\PrintMethodInfo\bin\Debug\netcoreapp3.1\PrintMethodInfo.dll
BaseSize:        0x18
ComponentSize:   0x0
Slots in VTable: 6
Number of IFaces in IFaceMap: 0
--------------------------------------
MethodDesc Table
           Entry       MethodDesc    JIT Name
00007FF99A080090 00007ff99a050a80   NONE System.Object.Finalize()
00007FF99A080098 00007ff99a050a90   NONE System.Object.ToString()
00007FF99A0800A0 00007ff99a050aa0   NONE System.Object.Equals(System.Object)
00007FF99A0800B8 00007ff99a050ae0   NONE System.Object.GetHashCode()
00007FF99A094CB0 00007ff99a131b50    JIT Program.Add(Int32, Int32)
00007FF99A092FC0 00007ff99a131b68    JIT Program.Main()
```

</details>

So it does seem that for .NET Core 3.1 it's something exclusive to being launched from Visual Studio.

**Comment by PathogenDavid** [comment]
Not sure how helpful this is, but I went spelunking in the runtime code a bit this morning to do some investigation. (I'm mostly leaving this comment for my own reference.)

ClrMd gets the `HotColdRegions` and the `CompilationType` from `ISOSDac::GetCodeHeaderData`. If I'm not horribly mistaken, this is eventually implemented by `ClrDataAccess::GetCodeHeaderData`. This method has basically two branches that result in success. [One in particular](https://github.com/dotnet/runtime/blob/6c1ae0da1382de25ce7ca498c927d4a2f87c4e12/src/coreclr/src/debug/daccess/request.cpp#L1245-L1250) is seemingly only supposed to happen when the method hasn't ben jitted. In this branch, everything will be 0 except for the `MethodDescPtr` (which as far as I can tell is the same thing as a method handle.)

ClrMd doesn't do anything with that field, but I made [a modified version of my test program](https://gist.github.com/PathogenDavid/cb5f3ee41fb6857c737bedec41f2e570) to query `ISOSDac` directly (thanks for leaving that accessible 🙏) and it does appear at a surface level that's the branch it's hitting:

```
.NET Core 3.1.9
Method handle: 0x00007FF9B49D1B50
Method: Program.Add(Int32, Int32)
Compilation type: None
Method start: 0x00007FF9B493A2B0
Cold region: 0x0000000000000000 for 0
Hot region: 0x00007FF9B493A2B0 for 0

Getting the CodeHeaderData directly:
======= Code header Data =======
GCInfo: 0x0000000000000000
JITType: 0
MethodDesc: 0x00007FF9B49D1B50
MethodStart: 0x0000000000000000
MethodSize: 0
ColdRegionStart: 0x0000000000000000
ColdRegionSize: 0
HotRegionSize: 0
Done.
```

Unfortunately my local runtime fork is hosed for some reason so I can't be entirely sure it's taking that branch. May have to investigate more later.
